### DISCUSSION TITLE
Voice Recorder

### DISCUSSION THREAD

**Discussion by ghost** [Boora-Raman/Voice_Recorder#1] [General]
Hii Have You Created Any Use Diagram Which We Can Use For Reference.

**Comment by ghost** [comment]
Can You Share That Use Case Diagram As Well As The Code Which You Were Showing To Me During Class.

**Reply by KhushiiChauhan** [reply]
yes the use case diagram is almost ready will share it soon

**Comment by KhushiiChauhan** [comment]
![Screenshot 2023-09-16 220905](https://github.com/Boora-Raman/Voice_Recorder/assets/142530032/387eeb9e-cf07-46c2-95cd-f1a891b44c55)
[](url)

**Reply by ghost** [reply]
Can You Also Commit The  Code  For The Above Use Case Diagram

**Comment by KhushiiChauhan** [comment]
i was working with some modification in code , this is the season of delay

**Reply by ghost** [reply]
Ok No Worries.

**Comment by ghost** [comment]
By The Way I Have Modified Our Use Case Diagram A Little Bit & I Have Also Created Pseudo Code For The Functionalities Which I Have Modified/Created I will Share The  Modified Diagram & Updated Code Once I Am Done With it.

**Comment by KhushiiChauhan** [comment]
gui is almost ready , prb is with play button

**Reply by ghost** [reply]
Ok I Have Created The Console Based Version Of It & Will Commit It Soon After That I Can Help You With The GUI Interface & Working For it.

**Comment by ghost** [comment]
Here Is The Latest Use Case Diagram Which I have Created & I have Added Some Additional Functionalities Other Than Those Which We Discussed Together & I Have Also Commited The Code For The Same.   
[Use_Case_diagram_Latest_Version.pdf](https://github.com/Boora-Raman/Voice_Recorder/files/12661946/Use_Case_diagram_Latest_Version.pdf)

**Reply by KhushiiChauhan** [reply]
nyc features added

**Comment by ghost** [comment]
Can You Share The GUI Code Which You Created So That We Can Debug It So That It Doesn't Produce Errors/Bugs

**Comment by KhushiiChauhan** [comment]
import sounddevice as sd
import numpy as np
import soundfile as sf
import tkinter as tk
from tkinter import messagebox

fs = 44100
recorded_data = []
stop_recording = False

input_stream = sd.InputStream(channels=2, samplerate=fs, dtype='float32')

# Function to start recording
def start_recording():
    global recorded_data, stop_recording
    recorded_data = []  # Reset the recorded data
    try:
        print("Recording Started. Press Stop to Stop Recording")
        while not stop_recording:
            audio_block, overflowed = input_stream.read(fs)
            recorded_data.extend(audio_block)
    except KeyboardInterrupt:
        pass

# Function to stop recording
def stop_recording():
    global stop_recording
    stop_recording = True
    input_stream.stop()
    input_stream.close()
    print("Recording stopped.")

# Function to play recorded audio
def play_recorded_audio():
    if recorded_data:
        print("Playing the recorded audio...")
        sd.play(np.array(recorded_data), fs)
        sd.wait()
    else:
        print("No data to play.")

# Create the main window
root = tk.Tk()
root.title("Audio Recorder")

# Create and configure the buttons
record_button = tk.Button(root, text="Record", command=start_recording)
stop_button = tk.Button(root, text="Stop", command=stop_recording)
play_button = tk.Button(root, text="Play", command=play_recorded_audio)
exit_button = tk.Button(root, text="Exit", command=root.quit)

# Pack the buttons into the window
record_button.pack()
stop_button.pack()
play_button.pack()
exit_button.pack()

# Start the input stream
input_stream.start()

# Run the GUI main loop
root.mainloop()
this is the basic model of gui of voice recoder

**Reply by KhushiiChauhan** [reply]
there is issue with play button.

**Reply by ghost** [reply]
Ok Will Try To Debug The Code

**Comment by KhushiiChauhan** [comment]
from tkinter import filedialog, messagebox, simpledialog
import tkinter as tk
import sounddevice as sd
import numpy as np
import soundfile as sf
import os

fs_low = 22050
fs_medium = 44100
fs_high = 96000
admin_username = "admin"
admin_password = "admin123"
audio_quality = 'medium'

fs_dict = {
    'low': fs_low,
    'medium': fs_medium,
    'high': fs_high
}

fs = fs_dict[audio_quality]

recorded_data = []

class VoiceRecorderApp:
    def _init_(self, root):
        self.root = root
        self.root.title("Voice Recorder App")
        self.stream = sd.InputStream(channels=2, samplerate=fs, callback=self.audio_callback)
        self.admin_mode = False  # Initialize admin_mode attribute
        self.folder_path = ''
        self.filename = ''
        self.status_bar = tk.Label(root, text="Idle", bd=1, relief=tk.SUNKEN, anchor=tk.W)
        self.status_bar.pack(side=tk.BOTTOM, fill=tk.X)

        self.start_button = tk.Button(root, text="Start Recording", command=self.start_recording)
        self.start_button.pack()

        self.stop_button = tk.Button(root, text="Stop Recording", command=self.stop_recording)
        self.stop_button.pack()

        self.cont_button = tk.Button(root, text="Continue Recording", command=self.continue_recording)
        self.cont_button.pack()

        self.play_button = tk.Button(root, text="Play Recorded Audio", command=self.play_recorded_audio)
        self.play_button.pack()

        self.save_button = tk.Button(root, text="Save Recorded Audio", command=self.save_audio)
        self.save_button.pack()

        self.delete_button = tk.Button(root, text="Delete File", command=self.delete_file)
        self.delete_button.pack()

        self.set_quality_button = tk.Button(root, text="Set Audio Quality", command=self.set_audio_quality)
        self.set_quality_button.pack()

        self.admin_login_button = tk.Button(root, text="Login to Admin Mode", command=self.login_to_admin_mode)
        self.admin_login_button.pack()

    def audio_callback(self, indata, frames, time, status):
        recorded_data.extend(indata.copy())

    def start_recording(self):
        global recorded_data
        recorded_data = []  # Reset recorded data
        self.stream.start()
        self.status_bar.config(text="Recording Started")

    def continue_recording(self):
        global recorded_data
        # recorded_data = []  # Reset recorded data
        self.stream.start()
        self.status_bar.config(text="Recording Continued")

    def stop_recording(self):
        self.stream.stop()
        # Do not close the stream, so it can be restarted
        self.status_bar.config(text="Recording Stopped")

    def play_recorded_audio(self):
        if recorded_data:
            sd.play(np.array(recorded_data), fs)
            sd.wait()
        else:
            messagebox.showwarning("Warning", "No data to play.")

    def set_audio_quality(self):
        global fs, audio_quality
        if self.admin_mode:
            new_quality = simpledialog.askstring("Audio Quality", "Enter the desired audio quality (low, medium, high):")
            if new_quality and new_quality.lower() in ['low', 'medium', 'high']:
                audio_quality = new_quality.lower()
                fs = fs_dict[audio_quality]
                messagebox.showinfo("Success", f"Audio quality set to {audio_quality}.")
            else:
                messagebox.showerror("Error", "Invalid quality option.")
        else:
            messagebox.showerror("Error", "Admin privileges required to set audio quality.")

    def login_to_admin_mode(self):
        global admin_username, admin_password
        username = simpledialog.askstring("Admin Login", "Enter username:")
        password = simpledialog.askstring("Admin Login", "Enter password:", show='*')
        if username == admin_username and password == admin_password:
            self.admin_mode = True
            messagebox.showinfo("Success", "Admin mode enabled.")
        else:
            messagebox.showerror("Error", "Invalid credentials.")

    def save_audio(self):
        global recorded_data
        if recorded_data:
            self.filename = filedialog.asksaveasfilename(defaultextension=".wav", initialdir=self.folder_path,
                                                        filetypes=[("WAV files", "*.wav")])
            if self.filename:
                sf.write(self.filename, np.array(recorded_data), fs)
                messagebox.showinfo("Success", f"Audio file saved as {os.path.basename(self.filename)} at {os.path.dirname(self.filename)}")
        else:
            messagebox.showwarning("Warning", "No data present to save.")

    def delete_file(self):
        if self.admin_mode:
            file_to_delete = filedialog.askopenfilename(initialdir=self.folder_path, title="Select file to delete",
                                                        filetypes=[("All files", "*.*")])
            if file_to_delete:
                try:
                    os.remove(file_to_delete)
                    messagebox.showinfo("Success", f"File deleted successfully at {os.path.dirname(file_to_delete)}")
                except FileNotFoundError:
                    messagebox.showerror("Error", "File not found.")
        else:
            messagebox.showwarning("Warning", "Admin Privaleges Required To Delete Files")            


root = tk.Tk()
app = VoiceRecorderApp(root)
root.mainloop()

**Reply by KhushiiChauhan** [reply]
this is the code as per we had discussed in the physical meet
